pkg/webhook/resourceusage/watcher.go

```diff
@@ -0,0 +1,182 @@
+	"k8s.io/client-go/tools/cache"
+	"sync"
+)
+
+type ResourceUsageWatcher struct {
+	currentUsageLock                     *sync.RWMutex
+	currentUsageByNamespace              map[string]*ResourceList
+	usageByNamespacePod                  map[string]map[string]*ResourceList
+	usageByNamespaceScheduledApplication map[string]map[string]*ResourceList
+	usageByNamespaceApplication          map[string]map[string]*ResourceList
+	crdInformerFactory                   crdinformers.SharedInformerFactory
+	coreV1InformerFactory                informers.SharedInformerFactory
+	podInformer                          corev1informers.PodInformer
+}
+
+// more convenient replacement for corev1.ResourceList
+type ResourceList struct {
+	cpu    resource.Quantity
+	memory resource.Quantity
+}
+
+const (
+	KindSparkApplication          = "SparkApplication"
+	KindScheduledSparkApplication = "ScheduledSparkApplication"
+)
+
+func (r ResourceList) String() string {
+	return fmt.Sprintf("cpu: %v mcore, memory %v bytes", r.cpu.MilliValue(), r.memory.Value())
+}
+
+func NewResourceUsageWatcher(crdInformerFactory crdinformers.SharedInformerFactory, coreV1InformerFactory informers.SharedInformerFactory) ResourceUsageWatcher {
+	glog.V(2).Infof("Creating new resource usage watcher")
+	r := ResourceUsageWatcher{
+		crdInformerFactory:                   crdInformerFactory,
+		currentUsageLock:                     &sync.RWMutex{},
+		coreV1InformerFactory:                coreV1InformerFactory,
+		currentUsageByNamespace:              make(map[string]*ResourceList),
+		usageByNamespacePod:                  make(map[string]map[string]*ResourceList),
+		usageByNamespaceScheduledApplication: make(map[string]map[string]*ResourceList),
+		usageByNamespaceApplication:          make(map[string]map[string]*ResourceList),
+	}
+	// Note: Events for each handler are processed serially, so no coordination is needed between
+	// the different callbacks. Coordination is still needed around updating the shared state.
+	sparkApplicationInformer := r.crdInformerFactory.Sparkoperator().V1beta1().SparkApplications()
+	sparkApplicationInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    r.onSparkApplicationAdded,
+		UpdateFunc: r.onSparkApplicationUpdated,
+		DeleteFunc: r.onSparkApplicationDeleted,
+	})
+	scheduledSparkApplicationInformer := r.crdInformerFactory.Sparkoperator().V1beta1().ScheduledSparkApplications()
+	scheduledSparkApplicationInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    r.onScheduledSparkApplicationAdded,
+		UpdateFunc: r.onScheduledSparkApplicationUpdated,
+		DeleteFunc: r.onScheduledSparkApplicationDeleted,
+	})
+	r.podInformer = r.coreV1InformerFactory.Core().V1().Pods()
+	r.podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    r.onPodAdded,
+		UpdateFunc: r.onPodUpdated,
+		DeleteFunc: r.onPodDeleted,
+	})
+	var podList []*corev1.Pod
```

Comment:
Can we put the code below for initially processing resource usages into a separate method?